pygatt/bled112_backend.py

```diff
@@ -0,0 +1,1977 @@
+            handler = logging.FileHandler(logfile)
+        else:  # print to stderr
+            handler = logging.StreamHandler()
+        formatter = logging.Formatter(fmt=LOG_FORMAT)
+        handler.setLevel(LOG_LEVEL)
+        handler.setFormatter(formatter)
+        self._logger.addHandler(handler)
+
+        # Initialization
+        self._lib = bled112_bglib.BGLib(loghandler=handler,
+                                        loglevel=LOG_LEVEL)
+        # Note: _ser is not protected by _main_thread_cond
+        self._ser = serial.Serial(serial_port, timeout=0.25)
+
+        # Main thread (the one calling commands) waits on this. This also
+        # provides mutual exclustion for BLED112Backend's state (except for the
+        # log)
+        self._main_thread_cond = threading.Condition()
+
+        # Packet receiving
+        self._recvr_thread = None  # background thread to receive packets
+        self._recvr_stop = True  # tell recvr thread not to run
+
+        # Flags to tell the main thread what to do when woken up from waiting
+        self._attribute_value_received = False  # attribute_value event occurred
+        self._bonded = False  # device is bonded
+        self._bond_expected = False  # tell bond_status handler to set _bonded
+        self._bonding_fail = False  # failed to bond with device
+        self._connect_timeout = False  # connection procedure timed-out
+        self._connected = False  # device is connected
+        self._encrypted = False  # connection is encrypted
+        self._event_return = 0  # event return code
+        self._procedure_completed = False  # procecure_completed event occurred
+        self._response_received = False  # command response received
+        self._response_return = 0  # command response return code
+
+        # BLED112_backend's state
+        self._attribute_value = None  # attribute_value event value
+        self._expected_attribute_handle = None  # expected handle after a read
+        self._bond_handle = 0xFF  # handle for the device bond
+        self._connection_handle = 0x00  # handle for the device connection
+        self._characteristics = {  # the device characteristics discovered
+            # uuid_string: Characteristic()
+        }
+        self._characteristics_cached = False  # characteristics already found
+        self._current_characteristic = None  # used in char/descriptor discovery
+        self._num_bonds = 0  # number of bonds stored on the BLED112
+        self._notifications = {  # stores notification packet contents
+            # handle: [value_bytearray0, ...]
+        }
+        self._stored_bonds = []  # bond handles stored on the BLED112
+        self._devices_discovered = {
+            # 'address': AdvertisingAndScanInfo,
+            # Note: address formatted like "01:23:45:67:89:AB"
+        }
+
+        # Register response handlers. Note: the packets of any response not
+        # registered here will go undetected.
```

Comment:
Because that's how the event handling system works with bled112_bglib. It allows any arbitrary function to be registered as a handler by using the += operator. The only alternative I can think of right now is to merge bled112_bglib with bled112_backend into one file and call the handlers directly.